scripts/correlation_test.py

```diff
@@ -117,30 +130,147 @@ def get_random_score_ratings(user_ratings_list, num):
         ans.append(permute_dict)
     return ans
 
+def sort(edge):
+    edge_list = sorted([e for e in edge])
+    result = (edge_list[0], edge_list[1])
+    return result
+
+def get_edge_set(user_ratings, graph):
+    '''
+    Create the set of birectional edges in the graph of users who have reviewed a particular restaurant.
+    user_ratings is a list of tuples (user, rating)
+    graph is a list of lists, the first entry in each list is a user, each subsequent entry is a friend of the user
+
+    return the set of bidirectional edges to be randomized
+    '''
+
+    #get the set of users in the user_ratings list
+    user_set = set()
+    for u in user_ratings:
+        user_set.add(u[0])
+
+    #create the original set of edges
+    edges = set()
+    #for each user that has reviewed this restaurant
+    for user in user_ratings:
+        start = user[0]    
+        #iterate over this user's friends     
+        for friend in graph[start]:
+            #check to see if this friend has reviewed the restaurant
+            if friend in user_set:
+                e = sort((start, friend))
+                edges.add(e)
+
+    return edges
+
+def is_valid_swap(edge_one, edge_two, edges):
+    '''
+    determine if edge two is a valid swap for edge one
+    edge_one and edge_two are tuples representing bidirectional edges
+    edges is the set of edges
+    '''
+    new_edge_one = sort((edge_two[0],edge_one[1]))
+    new_edge_two = sort((edge_one[0],edge_two[1]))
+
+    if new_edge_one in edges:
+        return False
+    if new_edge_two in edges:
+        return False
+
+    return True
+
+
+
+def get_random_edge_graphs(edges, num_permutations):
+    '''
+    Takes in the set of bidrectional edges representing the edges in the graph of users who have reviewed this restaurant
+    Return a list of user->list of friends dictionaries, each representing a randomized adjacency list
+    '''
+
+    #the list of randomized adjacency lists to return
+    randomized_graphs = []
+
+    for i in range(num_permutations):
+        randomized_edges = list(edges)
+        #create a copy of the edges set
+        e = copy.deepcopy(edges)
+        #iterate through each of the edges in the list
+        for edge in randomized_edges:
+            #if this edge has already been swapped out of the randomized set, skip it
+            if edge not in e:
+                continue
+            found_edge_to_swap = False
+            counter = 0
+            while found_edge_to_swap is False and counter < len(randomized_edges):
+                #somehow still getting numbers up to and including the upper bound, specification says this shouldn't be so
```

Comment:
like we shouldn't be performing len(randomized_edges) randomizations?

Reply:
so i see we upperbounded the number of times we try to swap the current edge we're looking at, but we're still not trying every possible edge because we're randomly selecting edges to try to swap with. It might be the case that we keep picking the same edge over and over again? Maybe we can try counter < len*2 or something

But also this might not be an issue if we just keep arbitrarily picking two edges from the set and attempt to swap them (some large number of times like the wiki article suggested?) I'm just slightly concerned that iterating through the edges in a particular order (and only doing each edge once) might introduce some nonrandom elements to the randomization result.